core/src/main/java/com/linecorp/armeria/common/stream/AbstractMulticastStreamFactory.java

```diff
@@ -0,0 +1,596 @@
+
+
+        private final Set<DownstreamSubscription> downstreamSubscriptions = new ConcurrentSet<>();
+
+        private final List<T> contentList = new ArrayList<>();
+
+        @SuppressWarnings("unused")
+        private volatile long requestedDemand;
+        private volatile Subscription upstreamSubscription;
+        private volatile boolean upstreamCompleted;
+        private volatile int upstreamOffset;
+        private volatile State state = State.OPEN;
+
+        private Throwable upstreamCause;
+
+        StreamMessageProcessor(StreamMessage<? extends T> upstream) {
+            upstream.subscribe(this, true);
+            notifyDownstreamWhenUpstreamCompleted(upstream);
+        }
+
+        private void notifyDownstreamWhenUpstreamCompleted(StreamMessage<? extends T> upstream) {
+            upstream.closeFuture().whenComplete((unused, cause) -> {
+                upstreamCompleted = true;
+                if (cause != null) {
+                    upstreamCause = cause;
+                }
+                notifyDownstreams();
+            });
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            upstreamSubscription = s;
+            downstreamSubscriptions.forEach(downstream -> {
+                if (downstream.updateOnSubscribe()) {
+                    if (downstream.executor() != null) {
+                        downstream.executor().execute(() -> downstream.subscriber().onSubscribe(downstream));
+                    } else {
+                        downstream.subscriber().onSubscribe(downstream);
+                    }
+                }
+            });
+            if (state == State.CLOSED) {
+                s.cancel();
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (upstreamOffset >= Integer.MAX_VALUE) {
+                //TODO(minwoox) limit by the size of data not by the size of contentList.
+                upstreamSubscription.cancel();
+                throw new IllegalStateException("Published numbers of data is greater than Integer.MAX_VALUE");
+            }
+            contentList.add(t);
+            upstreamOffset++;
+            notifyDownstreams();
+        }
+
+        /**
+         * Handled by {@link #notifyDownstreamWhenUpstreamCompleted(StreamMessage)} instead.
+         */
+        @Override
+        public void onError(Throwable t) {}
+
+        /**
+         * Handled by {@link #notifyDownstreamWhenUpstreamCompleted(StreamMessage)} instead.
+         */
+        @Override
+        public void onComplete() {}
+
+        void subscribe(DownstreamSubscription subscription) {
+            if (state != State.CLOSED) {
+                downstreamSubscriptions.add(subscription);
+                if (state != State.OPEN) {
+                    downstreamSubscriptions.remove(subscription);
+                    throw new IllegalStateException("This factory has been closed.");
+                }
+            }
```

Comment:
need to throw an `IllegalStateException` in `else` as well. Probably better using a boolean flag?